File: java/src/main/java/io/temporal/samples/moneytransfer/workflows/AccountTransferWorkflowScenarios.java
package io.temporal.samples.moneytransfer.workflows;

import io.temporal.common.SearchAttributeKey;
import io.temporal.common.converter.EncodedValues;
import io.temporal.failure.ActivityFailure;
import io.temporal.failure.ApplicationFailure;
import io.temporal.samples.moneytransfer.activities.AccountTransferActivities;
import io.temporal.samples.moneytransfer.model.DepositResponse;
import io.temporal.samples.moneytransfer.model.TransferInput;
import io.temporal.samples.moneytransfer.model.TransferOutput;
import io.temporal.samples.moneytransfer.model.TransferStatus;
import io.temporal.workflow.DynamicWorkflow;
import io.temporal.workflow.Workflow;
import org.slf4j.Logger;

import java.time.Duration;

public class AccountTransferWorkflowScenarios implements DynamicWorkflow {

    private static final String BUG = "AccountTransferWorkflowRecoverableFailure";
    private static final String NEEDS_APPROVAL = "AccountTransferWorkflowHumanInLoop";
    private static final String ADVANCED_VISIBILITY = "AccountTransferWorkflowAdvancedVisibility";

    private static final Logger log = Workflow.getLogger(AccountTransferWorkflowScenarios.class);

    private static final SearchAttributeKey<String> WORKFLOW_STEP = SearchAttributeKey.forKeyword("Step");

    private final AccountTransferActivities activities = Workflow.newActivityStub(
            AccountTransferActivities.class,
            AccountTransferActivities.activityOptions
    );

    private int progress = 0;
    private String transferState = "starting";
    private DepositResponse depositResponse = new DepositResponse("");

    private int approvalTime = 30;
    private boolean approved = false;

    @Override
    public Object execute(EncodedValues args) {
        Workflow.registerListener(new AccountTransferDynamicListenerImpl());
        TransferInput input = args.get(0, TransferInput.class);
        String type = Workflow.getInfo().getWorkflowType();
        log.info("Dynamic Account Transfer workflow started, type = {}", type);
        String idempotencyKey = Workflow.randomUUID().toString();

        // Validate
        upsertStep("Validate", type);
        activities.validate(input);
        updateProgress(25, 1);

        if (NEEDS_APPROVAL.equals(type)) {
            log.info(
                    "Waiting on 'approveTransfer' Signal or Update for workflow ID: {}",
                    Workflow.getInfo().getWorkflowId()
            );
            updateProgress(30, 0, "waiting");

            // Wait for the approval signal for up to approvalTime
            boolean receivedSignal = Workflow.await(Duration.ofSeconds(approvalTime), () -> approved);

            // If the signal was not received within the timeout, fail the workflow
            if (!receivedSignal) {
                log.error(
                        "Approval not received within the {}-second time window: Failing the workflow.",
                        approvalTime
                );
                throw ApplicationFailure.newFailure(
                        "Approval not received within " + approvalTime + " seconds",
                        "ApprovalTimeout"
                );
            }
        }

        // Withdraw
        upsertStep("Withdraw", type);
        activities.withdraw(idempotencyKey, input.getAmount(), type);
        updateProgress(50, 3);

        if (BUG.equals(type)) {
            // Simulate bug
            throw new RuntimeException("Simulated bug - fix me!");
        }

        // Deposit
        upsertStep("Deposit", type);
        try {
            depositResponse = activities.deposit(idempotencyKey, input.getAmount(), type);
            updateProgress(75, 1);
        } catch (ActivityFailure e) {
            // if deposit fails in an unrecoverable way, rollback the withdrawal and fail the workflow
            log.info("Deposit failed unrecoverable error, reverting withdraw");

            // undoWithdraw activity (rollback)
            activities.undoWithdraw(input.getAmount());

            // return failure message
            String message = ((ApplicationFailure) e.getCause()).getOriginalMessage();
            throw ApplicationFailure.newNonRetryableFailure(message, "DepositFailed");
        }

        // Send Notification
        upsertStep("Send Notification", type);
        activities.sendNotification(input);
        updateProgress(100, 1, "finished");

        return new TransferOutput(depositResponse);
    }

    private void upsertStep(String step, String type) {
        if (ADVANCED_VISIBILITY.equals(type)) {
            log.info("Advanced visibility .. {}", step);
            Workflow.upsertTypedSearchAttributes(WORKFLOW_STEP.valueSet(step));
        }
    }

    private void updateProgress(int progress, int sleep) {
        updateProgress(progress, sleep, "running");
    }

    private void updateProgress(int progress, int sleep, String transferState) {
        if (sleep > 0) {
            Workflow.sleep(Duration.ofSeconds(sleep));
        }
        this.transferState = transferState;
        this.progress = progress;
    }

    class AccountTransferDynamicListenerImpl implements AccountTransferMessages {

        @Override
        public TransferStatus queryTransferStatus() {
            return new TransferStatus(progress, transferState, "", depositResponse, approvalTime);
        }

        @Override
        public void approveTransferSignal() {
            log.info("Approve Signal Received");

            if (transferState.equals("waiting")) {
                approved = true;
            } else {
                log.info("Signal not applied: Transfer is not waiting for approval.");
            }
        }

        @Override
        public String approveTransferUpdate() {
            log.info("Approve Update Validated: Approving Transfer");
            approved = true;
            return "successfully approved transfer";
        }

        @Override
        public void approveTransferUpdateValidator() {
            log.info("Approve Update Received: Validating");
            if (approved) {
                throw new IllegalStateException("Validation Failed: Transfer already approved");
            }
            if (!transferState.equals("waiting")) {
                throw new IllegalStateException("Validation Failed: Transfer doesn't require approval");
            }
        }
    }
}

File: python/account_transfer_workflow_scenarios.py
from __future__ import annotations
import asyncio
from typing import Sequence, Any

from datetime import timedelta

from temporalio import workflow
from temporalio.common import RawValue, SearchAttributeKey
from temporalio.exceptions import ApplicationError

from shared_objects import TransferInput, TransferOutput, TransferStatus, DepositResponse
from activities import AccountTransferActivities

@workflow.defn(dynamic=True)
class AccountTransferWorkflowScenarios:

    BUG = "AccountTransferWorkflowRecoverableFailure"
    NEEDS_APPROVAL = "AccountTransferWorkflowHumanInLoop"
    ADVANCED_VISIBILITY = "AccountTransferWorkflowAdvancedVisibility"

    WORKFLOW_STEP = SearchAttributeKey.for_keyword("Step")

    approvalTime = 30
    approved = False

    def __init__(self) -> None:
        self.progress = 0
        self.transferState = ""
        self.depositResponse = DepositResponse("")
        self.start_to_close_timeout = timedelta(seconds=5)
        self.retry_policy = AccountTransferActivities.retry_policy

    @workflow.run
    async def execute(self, args: Sequence[RawValue]) -> Any:
        input = workflow.payload_converter().from_payload(args[0].payload, TransferInput)
        self.workflow_type = workflow.info().workflow_type
        logger = workflow.logger
        logger.info(f"Dynamic Account Transfer Workflow started {self.workflow_type}, input = {input}")
        idempotencyKey = str(workflow.uuid4())

        # Validate
        self.upsertStep("Validate")
        await workflow.execute_activity(AccountTransferActivities.validate,
            args=[input],
            start_to_close_timeout=self.start_to_close_timeout,
            retry_policy=self.retry_policy)
        await self.updateProgress(25, 1)

        if self.NEEDS_APPROVAL == self.workflow_type:
            logger.info(f"Waiting on 'approveTransfer' Signal or Update for workflow ID: {workflow.info().workflow_id}")
            await self.updateProgressStatus(30, 0, "waiting")

            # Wait for approval for up to approvalTime
            try:
                await workflow.wait_condition(lambda: self.approved, timeout=timedelta(seconds=self.approvalTime))
            except asyncio.TimeoutError:
                logger.error(f"Approval not received within the {self.approvalTime} seconds")
                raise ApplicationError(f"Approval not received within {self.approvalTime} seconds", type="ApprovalTimeout", non_retryable=True)

        # Withdraw
        self.upsertStep("Withdraw")
        await workflow.execute_activity(AccountTransferActivities.withdraw,
            args=[idempotencyKey, input.amount, self.workflow_type],
            start_to_close_timeout=self.start_to_close_timeout,
            retry_policy=self.retry_policy)
        await self.updateProgress(50, 3)

        if self.BUG == self.workflow_type:
            # simulate bug
            raise RuntimeError("Simulated bug - fix me!")

        # Deposit
        self.upsertStep("Deposit")
        try:
            self.depositResponse = await workflow.execute_activity(AccountTransferActivities.deposit,
                args=[idempotencyKey, input.amount, self.workflow_type],
                start_to_close_timeout=self.start_to_close_timeout,
                retry_policy=self.retry_policy)
            await self.updateProgress(75, 1)
        except Exception as ex:
            logger.info(f"Depsoit failed unrecoverable error, reverting withdraw")
            # Undo Withdraw (rollback)
            await workflow.execute_activity(AccountTransferActivities.undoWithdraw,
                args=[input.amount],
                start_to_close_timeout=self.start_to_close_timeout,
                retry_policy=self.retry_policy)

            # return failure message
            raise ApplicationError(f"{ex.__cause__} ", type="DepositFailed", non_retryable=True)

        # Send Notification
        self.upsertStep("Send notification")
        await workflow.execute_activity(AccountTransferActivities.sendNotification,
            args=[input],
            start_to_close_timeout=self.start_to_close_timeout,
            retry_policy=self.retry_policy)
        await self.updateProgressStatus(100, 1, "finished")

        return TransferOutput(DepositResponse(self.depositResponse.get_chargeId()))

    @workflow.signal(name="approveTransfer")
    def approveTransferSignal(self) -> str:
        workflow.logger.info(f"Approve Signal Received")
        if self.transferState == "waiting":
            self.approved = True
        else:
            workflow.logger.info(f"Approval not applied. Transfer is not waiting for approval")

    # TODO: Implement Update
    def approveTransferUpdate(self) -> str:
        workflow.logger.info(f"Approve Update Validated: Approving Transfer")
        self.approved = True
        return "successfully approved transfer"

    # TODO: Implement Validator
    def approveTransferUpdateValidator(self) -> None:
        workflow.logger.info(f"Approve Update Received: Validating")
        if approved:
            # TODO: Validate this is the right error to throw
            raise ApplicationError(f"Validation Failed: Transfer already approved")
        if self.transferState != "waiting":
            # TODO: validate this is the right error to throw
            raise ApplicationError(f"Validation Failed: Transfer doesn't require approval")


    @workflow.query(name="transferStatus")
    def queryTransferStatus(self) -> TransferStatus:
        return TransferStatus(self.progress, self.transferState, "", self.depositResponse, self.approvalTime)

    def upsertStep(self, step: str) -> None:
        if self.ADVANCED_VISIBILITY == self.workflow_type:
            workflow.logger.info(f"Advanced visibilty... On step: {step}")
            workflow.upsert_search_attributes([self.WORKFLOW_STEP.value_set(step)])


    async def updateProgress(self, progress: int, sleep: int) -> None:
        await self.updateProgressStatus(progress, sleep, "running")

    async def updateProgressStatus(self, progress: int, sleep: int, transferState: str) -> None:
        if sleep > 0:
            await asyncio.sleep(sleep)

        self.transferState = transferState
        self.progress = progress
